File: module-app/src/main/java/maple/expectation/monitoring/collector/RedisMetricsCollector.java
package maple.expectation.monitoring.collector;

import io.micrometer.core.instrument.Gauge;
import io.micrometer.core.instrument.MeterRegistry;
import java.util.LinkedHashMap;
import java.util.Map;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import maple.expectation.domain.repository.RedisBufferRepository;
import maple.expectation.infrastructure.config.MonitoringThresholdProperties;
import org.springframework.stereotype.Component;

/**
 * Redis 메트릭 수집기 (Issue #251)
 *
 * <h3>수집 항목</h3>
 *
 * <ul>
 *   <li>버퍼 대기 항목 수
 *   <li>캐시 히트율
 *   <li>커넥션 상태
 * </ul>
 *
 * @see MetricsCollectorStrategy
 */
@Slf4j
@Component
@RequiredArgsConstructor
public class RedisMetricsCollector implements MetricsCollectorStrategy {

  private final MeterRegistry meterRegistry;
  private final RedisBufferRepository redisBufferRepository;
  private final MonitoringThresholdProperties thresholdProperties;

  @Override
  public String getCategoryName() {
    return MetricCategory.REDIS.getKey();
  }

  @Override
  public Map<String, Object> collect() {
    Map<String, Object> metrics = new LinkedHashMap<>();

    // 버퍼 상태
    collectBufferMetrics(metrics);

    // 캐시 히트율
    collectCacheMetrics(metrics);

    return metrics;
  }

  @Override
  public boolean supports(MetricCategory category) {
    return MetricCategory.REDIS == category;
  }

  @Override
  public int getOrder() {
    return 5;
  }

  private void collectBufferMetrics(Map<String, Object> metrics) {
    // Redis 버퍼 대기 항목 수 (직접 조회)
    long pendingCount = redisBufferRepository.getTotalPendingCount();
    metrics.put("buffer_pending_count", pendingCount);

    // 버퍼 포화도 (임계값 기준 - 설정에서 가져옴)
    double saturation = (pendingCount / thresholdProperties.getBufferSaturationDouble()) * 100;
    metrics.put("buffer_saturation_percent", Math.min(formatDouble(saturation), 100.0));

    // Micrometer 메트릭에서 추가 정보
    Gauge bufferGauge = meterRegistry.find("redis.buffer.pending").gauge();
    if (bufferGauge != null) {
      metrics.put("buffer_gauge_value", (long) bufferGauge.value());
    }
  }

  private void collectCacheMetrics(Map<String, Object> metrics) {
    // L1 캐시 (Caffeine) 히트율
    Gauge caffeineHits = meterRegistry.find("cache.gets").tag("result", "hit").gauge();
    Gauge caffeineMisses = meterRegistry.find("cache.gets").tag("result", "miss").gauge();

    if (caffeineHits != null && caffeineMisses != null) {
      double hits = caffeineHits.value();
      double misses = caffeineMisses.value();
      double total = hits + misses;
      if (total > 0) {
        metrics.put("l1_cache_hit_rate", formatDouble((hits / total) * 100));
      }
    }

    // L2 캐시 (Redis) 관련 메트릭
    var l2HitCounter = meterRegistry.find("tiered.cache.hit").tag("layer", "L2").counter();
    var l2MissCounter = meterRegistry.find("tiered.cache.miss").tag("layer", "L2").counter();

    if (l2HitCounter != null && l2MissCounter != null) {
      double l2Hits = l2HitCounter.count();
      double l2Misses = l2MissCounter.count();
      double l2Total = l2Hits + l2Misses;
      if (l2Total > 0) {
        metrics.put("l2_cache_hit_rate", formatDouble((l2Hits / l2Total) * 100));
      }
    }
  }

  private double formatDouble(double value) {
    if (Double.isNaN(value) || Double.isInfinite(value)) {
      return 0.0;
    }
    return Math.round(value * 100.0) / 100.0;
  }
}

File: module-infra/src/main/kotlin/maple/expectation/infrastructure/monitoring/collector/RedisMetricsCollector.kt
package maple.expectation.infrastructure.monitoring.collector

import io.micrometer.core.instrument.Counter
import io.micrometer.core.instrument.Gauge
import io.micrometer.core.instrument.MeterRegistry
import maple.expectation.core.port.out.BufferStatusQuery
import maple.expectation.infrastructure.config.MonitoringThresholdProperties
import org.springframework.stereotype.Component
import kotlin.math.min
import kotlin.math.round

/**
 * Redis 메트릭 수집기
 *
 * <p>버퍼 상태와 캐시 성능 메트릭을 수집합니다.
 *
 * <h2>DIP 준수</h2>
 *
 * <p>{@link BufferStatusQuery} Port를 통해 버퍼 상태를 조회하므로 Repository 직접 참조를 방지합니다.
 *
 * @see BufferStatusQuery 버퍼 상태 조회 Port
 */
@Component
class RedisMetricsCollector(
    private val meterRegistry: MeterRegistry,
    private val bufferStatus: BufferStatusQuery,
    private val thresholdProperties: MonitoringThresholdProperties
) : MetricsCollectorStrategy {

  override fun getCategoryName(): String = MetricCategory.REDIS.key

  override fun collect(): Map<String, Any> = buildMap {
    collectBufferMetrics(this)
    collectCacheMetrics(this)
  }

  override fun supports(category: MetricCategory): Boolean =
      MetricCategory.REDIS == category

  override fun getOrder(): Int = 5

  private fun collectBufferMetrics(metrics: MutableMap<String, Any>) {
    val pendingCount = bufferStatus.getTotalPendingCount()
    metrics["buffer_pending_count"] = pendingCount

    val saturation = (pendingCount.toDouble() / thresholdProperties.bufferSaturationDouble) * 100
    metrics["buffer_saturation_percent"] = min(formatDouble(saturation), 100.0)

    val bufferGauge = meterRegistry.find("redis.buffer.pending").gauge()
    bufferGauge?.let {
      metrics["buffer_gauge_value"] = it.value().toLong()
    }
  }

  private fun collectCacheMetrics(metrics: MutableMap<String, Any>) {
    val caffeineHits = meterRegistry.find("cache.gets").tag("result", "hit").gauge()
    val caffeineMisses = meterRegistry.find("cache.gets").tag("result", "miss").gauge()

    if (caffeineHits != null && caffeineMisses != null) {
      val hits = caffeineHits.value()
      val misses = caffeineMisses.value()
      val total = hits + misses
      if (total > 0) {
        metrics["l1_cache_hit_rate"] = formatDouble((hits / total) * 100)
      }
    }

    val l2HitCounter = meterRegistry.find("tiered.cache.hit").tag("layer", "L2").counter()
    val l2MissCounter = meterRegistry.find("tiered.cache.miss").tag("layer", "L2").counter()

    if (l2HitCounter != null && l2MissCounter != null) {
      val l2Hits = l2HitCounter.count()
      val l2Misses = l2MissCounter.count()
      val l2Total = l2Hits + l2Misses
      if (l2Total > 0) {
        metrics["l2_cache_hit_rate"] = formatDouble((l2Hits / l2Total) * 100)
      }
    }
  }

  private fun formatDouble(value: Double): Double {
    if (value.isNaN() || value.isInfinite()) {
      return 0.0
    }
    return round(value * 100.0) / 100.0
  }
}
